### PR TITLE
Feature/parsing counting

### PR DESCRIPTION
**Pull request by kuchichan**
Ready to review

### CODE REVIEW

File: bowling/inc/frameParser.hpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <string>
+#include <vector>
+
+class FrameParser {
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp:7]: In my opinion FrameParser shoud have static function to parse frameString and return vector of frames.

File: bowling/inc/frameParser.hpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+public:
+    FrameParser(const std::string& gameString);  // thinking about explicit...
+
+    std::string getGameString() const { return gameString_; }
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: redundant

File: bowling/inc/frameParser.hpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    std::vector<size_t> getFrames() const { return frames_; }
+
+    void parse();
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: can be turn into  static function
```suggestion
    static std::vector<size_t> parse(const std::string& gamePoints){
        //parsingString into frames vector
        return frames;
    }
```

File: bowling/inc/frameParser.hpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+#include <string>
+#include <vector>
+
+class FrameParser {
+public:
+    FrameParser(const std::string& gameString);  // thinking about explicit...
+
+    std::string getGameString() const { return gameString_; }
+    std::vector<size_t> getFrames() const { return frames_; }
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: redundant

File: bowling/inc/frameParser.hpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    void parse();
+
+private:
+    std::string gameString_;
+    std::vector<size_t> frames_;
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: redundant

File: bowling/inc/game.hpp [hunk 6]
```diff
@@ -2,10 +2,10 @@
...
 #include <iostream>
 
-constexpr int MAX_PINS{ 10 };
-constexpr int MAX_FRAMES{ 10 };
-constexpr int MAX_ROLLS{ 21};
-constexpr int MAX_FRAME_SCORE{ MAX_PINS };
+constexpr int MAX_PINS{10};
+constexpr int MAX_FRAMES{10};
+constexpr int MAX_ROLLS{21};
+constexpr int MAX_FRAME_SCORE{MAX_PINS};
```
Comment (kszytko) [inline bowling/inc/game.hpp:8]: do not modify other class if you work only on frameParser.  Make it different PR

File: bowling/src/main.cpp [hunk 7]
```diff
@@ -1,5 +1,3 @@
 #include "game.hpp"
 
-int main() {
-
-}
+int main() {}
```
Comment (kszytko) [inline bowling/src/main.cpp]: same up

File: bowling/src/game.cpp [hunk 8]
```diff
@@ -9,12 +9,10 @@ int Game::score() {
     int firstFrameThrow{};
     for (int frame = 0; frame < MAX_FRAMES; ++frame) {
         if (isStrike(firstFrameThrow)) {
-            score_ += MAX_FRAME_SCORE +
-                      nextTwoBallsForStrike(firstFrameThrow);
+            score_ += MAX_FRAME_SCORE + nextTwoBallsForStrike(firstFrameThrow);
```
Comment (kszytko) [inline bowling/src/game.cpp:12]: same up

**Review by kszytko — CHANGES_REQUESTED** [review]
Some of ideas to make. Modifiyng formating in other classes should go into other PR

File: bowling/inc/frameParser.hpp (hunk 1 above)
Comment (kuchichan) [reply bowling/inc/frameParser.hpp:7]: We will see.

File: bowling/inc/frameParser.hpp (hunk 3 above)
Comment (kuchichan) [reply bowling/inc/frameParser.hpp]: This is vip, we will see how it goes.

File: bowling/inc/game.hpp (hunk 6 above)
Comment (kuchichan) [reply bowling/inc/game.hpp:8]: Agreed, done.

File: bowling/inc/frameParser.hpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include <string>
+#include <vector>
+
+constexpr char DELIMITER = '|';
+constexpr char STRIKE = 'X';  // Maybe map like - strike - 10, spare miss, 0, but additional logic with spare sign...
+constexpr char SPARE_SIGN = '/';
+constexpr char MISS = '-';
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: move it into class,   only class need it.    
Minus is that constexpr doesn't work in classes, instead use normal const.

File: bowling/inc/frameParser.hpp [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+using tokenizedFrame = std::pair<std::string,  std::vector<char>>;
+using parsedFrame = std::pair<std::string, std::vector<size_t>>;
+
+class FrameParser {
+public:
+    static parsedFrame parse(const std::string& gamePoints);
+};
+
+class Tokenizer {
+public:
```
Comment (kszytko) [inline bowling/inc/frameParser.hpp]: i dont think this class in needed

File: bowling/src/frameParser.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+		     last_parsed_char = 0;
+		     return last_parsed_char;
+		   } else {
+		     last_parsed_char = el - '0';
+		     return last_parsed_char;
+		   }
+		 });
+  
+  return std::pair(tokenized.first, parsed);
+}
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: to long function,  split to smallest private methods

File: bowling/src/frameParser.cpp [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+		     last_parsed_char = 10 - last_parsed_char;
+		     return last_parsed_char;
+		   } else if (el == '-') {
+		     last_parsed_char = 0;
+		     return last_parsed_char;
+		   } else {
+		     last_parsed_char = el - '0';
+		     return last_parsed_char;
+		   }
+		 });
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: too big lambda function

File: bowling/src/frameParser.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+#include <iostream>
+#include <iterator>
+#include <algorithm>
+
+parsedFrame FrameParser::parse(const std::string& gamePoints) {
+  std::vector<size_t> parsed {};
+  size_t last_parsed_char{};
+  auto tokenized = Tokenizer::splitToFramesAlt(gamePoints);
+  std::transform(tokenized.second.cbegin(),
+		 tokenized.second.cend(),
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: bad idea to modify item of pair like this. very bad

File: bowling/src/frameParser.cpp [hunk 14]
```diff
@@ -0,0 +1,54 @@
+#include "frameParser.hpp"
+
+#include <iostream>
+#include <iterator>
+#include <algorithm>
+
+parsedFrame FrameParser::parse(const std::string& gamePoints) {
+  std::vector<size_t> parsed {};
+  size_t last_parsed_char{};
+  auto tokenized = Tokenizer::splitToFramesAlt(gamePoints);
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: hmm,  maybe better name should be  pointList ? parsedList?   what exaclty splitToFramesAlt does?

File: bowling/src/frameParser.cpp [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+  std::transform(tokenized.second.cbegin(),
+		 tokenized.second.cend(),
+		 back_inserter(parsed),
+		 [&last_parsed_char] (char el) {
+		   if (el == 'X') {
+		     last_parsed_char = 10;
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: we use camelCase variable convention

Comment (kszytko) [reply bowling/src/frameParser.cpp]: what exacly last_parsed_char hold?   why not char?  name does not tell us anything

File: bowling/src/frameParser.cpp [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    next = gamePoints.find(":", prev);
+    
+    if (next != std::string::npos) {
+      name = gamePoints.substr(prev, next - prev);
+      prev = next + 1;
+    }
+    
+    std::vector<char> result = {
+        std::next(gamePoints.begin(), prev),
+        gamePoints.end()};
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: you did something, but it's not readable totally

File: bowling/src/frameParser.cpp [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    
+    result.erase(std::remove_if(result.begin(), result.end(),
+				[] (char el) {return el == '|'; }),
+		 result.end());
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]:
```suggestion
    auto separator = '|';  //or make it const
    result.erase(std::remove(result.begin(), result.end(),  separator ),
		 result.end());
```

File: bowling/src/frameParser.cpp [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    size_t prev = 0;
+    size_t next = 0;
+    std::string name = "";
+    
+    next = gamePoints.find(":", prev);
+    
+    if (next != std::string::npos) {
+      name = gamePoints.substr(prev, next - prev);
+      prev = next + 1;
+    }
```
Comment (kszytko) [inline bowling/src/frameParser.cpp]: Separate into 2 methods, one for readName, second for matchSequence

Comment (kszytko) [reply bowling/src/frameParser.cpp]: Question!  What type of data will you get from folderReader class?   lineString without parsed name?

File: bowling/src/frameParser.cpp (hunk 14 above)
Comment (kszytko) [inline bowling/src/frameParser.cpp]: Separate into 2 methods, f.e. one for readName, second for readSequence:
```suggestion
  auto name= readName(line);
  auto sequence = readSequence(line);
```

Comment (kszytko) [reply bowling/src/frameParser.cpp]: split it into two methods something like: 
```suggestion
  auto playerName = readPlayerName(gameLine);
  auto gameSequence = readSequence(gameLine);
```
you dont need to use pairs

**Review by kszytko — CHANGES_REQUESTED** [review]
Some changes,  please write tests.

File: bowling/src/frameParser.cpp (hunk 17 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: I will fix all these magic chars (look constexpr -> const in class)

File: bowling/src/frameParser.cpp (hunk 16 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Why? I Just use the fact, that vector can take iterator ranges, and this is efficient way to switch from string to vector of chars.

File: bowling/src/frameParser.cpp (hunk 15 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: last_parser_char holds last calculated "shot " and its use to calculate exact value from "/" -> that is the way that is used in game.cpp

File: bowling/src/frameParser.cpp (hunk 13 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Could you elaborate, why? Maybe some source, materials. Im thinking about leaving pairs at all in favour of just simple vector where first element will be the name - im just little bit worried for example about strings without names.

File: bowling/src/frameParser.cpp (hunk 16 above)
Comment (kszytko) [reply bowling/src/frameParser.cpp]: String is a vector of char itself

File: bowling/src/frameParser.cpp (hunk 13 above)
Comment (kszytko) [reply bowling/src/frameParser.cpp]: Becouse as reader i dont know what this pair holds,  better way is to use separate variables for name and array

File: bowling/src/frameParser.cpp (hunk 12 above)
Comment (kszytko) [reply bowling/src/frameParser.cpp]: Extract it into method

File: bowling/inc/frameParser.hpp (hunk 9 above)
Comment (kszytko) [reply bowling/inc/frameParser.hpp]: And consider to use or not `_SIGN` in every const

Comment (kuchichan) [reply bowling/inc/frameParser.hpp]: Done.

File: bowling/inc/frameParser.hpp (hunk 10 above)
Comment (kuchichan) [reply bowling/inc/frameParser.hpp]: I think it is needed, since its easier to test and we are not implementing antipattern "iceberg" class

File: bowling/src/frameParser.cpp (hunk 11 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 12 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 13 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 15 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Made somewhat easier to read.

File: bowling/src/frameParser.cpp (hunk 16 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: That's right. Done.

File: bowling/src/frameParser.cpp (hunk 17 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 18 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: I do not know ;p I assumed, that i have to parse all. Due to your proposals /advice its easy to turn this part off :)

File: bowling/src/frameParser.cpp (hunk 14 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+        lastParsedToken = 10 - lastParsedToken;
+        return lastParsedToken;
+    } else if (token == FrameParser::MissSign) {
+        lastParsedToken = 0;
+        return lastParsedToken;
+    } else {
+        lastParsedToken = token - '0';
+        return lastParsedToken;
+    }
+}
```
Comment (niemarcin) [inline bowling/src/frameParser.cpp]: 1. You don't need to `return lastParsedToken;` in every condition. Just do it at the end of the method.
2. Do you think switch-case instead of if-else could be better?

File: bowling/src/frameParser.cpp [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "frameParser.hpp"
+#include "tokenizer.hpp"
+
+const char FrameParser::StrikeSign = 'X';
+const char FrameParser::MissSign = '-';
+const char FrameParser::SpareSign = '/';
+
+
+size_t FrameParser::parseToken(const char token, size_t& lastParsedToken) {
```
Comment (niemarcin) [inline bowling/src/frameParser.cpp]: It seems inconsistent that `parseToken` method is used to parse a token which is `const char` and in the same line we have `lastParseToken` which type is `size_t&`. Both are tokens (current and previous) but have different types. 

Maybe the `lastParsedToken` name could be changed to something different? `lastTokenValue` or `previousTokenValue` or something like that.

File: bowling/src/frameParser.cpp [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        lastParsedToken = 10;
+        return lastParsedToken;
+    } else if (token == FrameParser::SpareSign) {
+        lastParsedToken = 10 - lastParsedToken;
+        return lastParsedToken;
+    } else if (token == FrameParser::MissSign) {
+        lastParsedToken = 0;
+        return lastParsedToken;
+    } else {
+        lastParsedToken = token - '0';
```
Comment (niemarcin) [inline bowling/src/frameParser.cpp]: Explicite type cast might be better.

File: bowling/src/frameParser.cpp (hunk 20 above)
Comment (niemarcin) [reply bowling/src/frameParser.cpp]: Another thing - I don't know if it it's a good practise to return the same value by `return` statement from function and at the same time through a reference to parameter. I can understand why it's used like this by looking at  `parse()` method, but I'm just wondering if it's good or bad to use it like this.

File: bowling/src/tokenizer.cpp [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+
+std::string Tokenizer::readName(const std::string& line, size_t& position) {
+    size_t next = line.find(NameDelimiterSign);
+    if (next != std::string::npos) {
+        position = next;
+        return line.substr(0, next);
+    } else {
+        return "";
+    }
+}
```
Comment (niemarcin) [inline bowling/src/tokenizer.cpp]: It took me a while to understand how these functions work. 

I don't know if it's a good idea to make it like this. Now `readSequence()` and `readName()` methods are tightly coupled by the `position` parameters and the way how they are called in FrameParser class. Wouldn't be better to just remove `position` from a parameters and once again use `line.find(NameDelimiterSign)` ? I know that it's less efficient but readability and **independence** (no coupling between) of these methods will be better.

So it would be something like this:
```cpp
std::string Tokenizer::readSequence(const std::string& line) {
    size_t position = line.find(NameDelimiterSign);
    if (position == std::string::npos) {
        return "";
    }
    std::string result = {std::next(line.begin() + position), line.end()};
    result.erase(std::remove(result.begin(), result.end(), FrameDelimiterSign), result.end());
    return result;
}

std::string Tokenizer::readName(const std::string& line) {
    size_t position = line.find(NameDelimiterSign);
    if (position != std::string::npos) {
        return line.substr(0, position);
    } else {
        return "";
    }
}
```

And then in `FrameParser::parse`  the `size_t currentLinePosition{};` is not needed.

And of course tokenizer-ut will change
```cpp
TEST_P(GetNameSuite, TestReadNameReturnsPartOfTheStringToColonSign) {
    auto [param, expected] = GetParam();
    ASSERT_EQ(Tokenizer::readName(param), expected);
}

TEST_P(GetSeqSuite, TestReadSequenceReturnsStringWithoutNameAndDelimiter) {
    auto [param, expected] = GetParam();
    ASSERT_EQ(Tokenizer::readSequence(param), expected);
}
```

Or maybe we can find the `NameDelimiterSign` position once and store it somewhere for future use? But finding this delimiter should **not** be done in `readName` method for later use in `readSequence`. Maybe additional `findNameDelimiter()` method should be created?

**Review by niemarcin — CHANGES_REQUESTED** [review]
I don't get one thing in general - why both `FrameParser `and `Tokenizer `classes are fully `static`?

File: bowling/src/frameParser.cpp (hunk 20 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: To be honest, I have also feeling that it is not the best way to do that. I have tried with lambda function and capture list, but this was also derided in previous review. I do not really have idea what to do. I will need this last parsed element anyway. 

To the first comment - please note that is actually "calculated" value (integer/size_t) of token 10, 5 ,3 or so, not token itself '/' '3'.

File: bowling/src/frameParser.cpp (hunk 21 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Could you be more specific? Like static_cast<size_t> ? This will not produce needed output.

File: bowling/src/tokenizer.cpp (hunk 22 above)
Comment (kuchichan) [reply bowling/src/tokenizer.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 19 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 20 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: I will leave as it is (maybe there is obvoius way for doing this - there kind of problems are probably common).

File: bowling/src/frameParser.cpp (hunk 21 above)
Comment (niemarcin) [reply bowling/src/frameParser.cpp]: Why not? I've just checked `lastParsedToken = static_cast<size_t>(token - '0');` - it compiles and tests still pass.

I'm not sure if it's better to use implicit or explicit casting, but based on experience with other languages I believe it's better to avoid implicit type conversion.

File: bowling/src/frameParser.cpp (hunk 20 above)
Comment (niemarcin) [reply bowling/src/frameParser.cpp]: > To the first comment - please note that is actually "calculated" value (integer/size_t) of token 10, 5 ,3 or so, not token itself '/' '3'.

Then maybe `lastRollValue` or just `lastValue` ?

Comment (niemarcin) [reply bowling/src/frameParser.cpp]: And related to other stuff:

I've tried this:
```cpp
void FrameParser::parseToken(const char token, size_t& lastParsedToken) {
    switch (token) {
    case FrameParser::StrikeSign:
        lastParsedToken = 10;
        break;
    case FrameParser::SpareSign:
        lastParsedToken = 10 - lastParsedToken;
        break;
    case FrameParser::MissSign:
        lastParsedToken = 0;
        break;
    default:
        lastParsedToken = static_cast<size_t>(token - '0');
        break;
    }
}
```
As you can see I've changed return type to `void`.

And in `FrameParser::parse()` method, instead of `std::transform` I've called this:
```cpp
    std::for_each(sequence.cbegin(), sequence.cend(),
                  [&lastParsedToken, &parsed](const auto& el){
                      FrameParser::parseToken(el, lastParsedToken);
                      parsed.push_back(lastParsedToken);
                  });
```

Just my suggestion, not the ultimate solution.  ;) But tests are still fine.

Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

File: bowling/src/frameParser.cpp (hunk 21 above)
Comment (kuchichan) [reply bowling/src/frameParser.cpp]: Done.

**Review by niemarcin — APPROVED** [review]
Great work!

File: bowling/inc/frameParser.hpp (hunk 1 above)
Comment (kszytko) [reply bowling/inc/frameParser.hpp:7]: I was wrong,  static function is not good idea. We should operate here on normal object.

**Review by kszytko — CHANGES_REQUESTED** [review]
Parser not working with empty shoots:
```diff
- "1-||X "-> {1,0,10,0},  is
+ "1-||X "-> {1,0,0,0,10,0},   should be
```

File: bowling/inc/frameParser.hpp (hunk 1 above)
Comment (kuchichan) [reply bowling/inc/frameParser.hpp:7]: I will leave It as it is.

**Comment by kuchichan** [comment]
> Parser not working with empty shoots:
> 
> ```diff
> - "1-||X "-> {1,0,10,0},  is
> + "1-||X "-> {1,0,0,0,10,0},   should be
> ```

Except that is completely wrong input.

**Comment by kszytko** [comment]
> > Parser not working with empty shoots:
> > ```diff
> > - "1-||X "-> {1,0,10,0},  is
> > + "1-||X "-> {1,0,0,0,10,0},   should be
> > ```
> 
> Except that is completely wrong input.

Ok, my mistake.

**Review by kszytko — APPROVED** [review]
